### ISSUE TITLE
Dependency Resolver is too picky

### ISSUE THREAD

**Issue by LasneF** [siom79/japicmp#76]
hello 
it look like the methods setUpClassPathUsingMavenProject  is gathering too munch jars

    Set<Artifact> dependencyArtifacts = mavenParameters.getMavenProject().getArtifacts();
		for (Artifact artifact : dependencyArtifacts) {
			String scope = artifact.getScope();
			if (!"test".equals(scope)) {
                                ///  Is that mandatory  ?  <<==
				Set<Artifact> artifacts = resolveArtifact(artifact, mavenParameters, true);
				for (Artifact resolvedArtifact : artifacts) {
					File resolvedFile = resolvedArtifact.getFile();
					if (getLog().isDebugEnabled()) {
						getLog().debug("Adding to classpath: " + resolvedFile.getAbsolutePath() + "; scope: " + scope);
					}
					comparatorOptions.getClassPathEntries().add(resolvedFile.getAbsolutePath());
				}
			}
		}

the I do not get the usage of the second loop as the 1st level dependency tree is already the full maven dependency tree   like maven-dependency-plugin:2.10:tree would have produced 

for instance on my project i get in my pom i get 12 entries . The dependencyArtifacts variables gets already my 49 entries ,(as same as maven dependency tree) 
but resolving artifacts on those 49 gives 169 artifacts ! 

and in fact this methods ask for jar that i even do not have in my maven repo ... and crashes 

i wonder if it is a bug or something usefull. 

Thx for your tools 

NB: if i get time i would like to do a Sonar pluggin like it exists for  
http://docs.sonarqube.org/display/PLUG/Clirr+Plugin

**Comment by siom79** [comment]
Can you provide a concrete example that illustrates the difference?

When I add for example the following dependency `com.fasterxml.jackson.core:jackson-databind:jar:2.6.3:compile` to a project, then the output of the dependency plugin is that:
```
com.github.siom79.japicmp:japicmp-test-maven-plugin-jackson:jar:0.6.1-SNAPSHOT
+- com.fasterxml.jackson.core:jackson-databind:jar:2.6.3:compile
|  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.6.0:compile
|  \- com.fasterxml.jackson.core:jackson-core:jar:2.6.3:compile
```
The method `setUpClassPathUsingMavenProject()` you have cited above returns the same three dependencies.
Although they are really added twice, hence I have created the new branch `issue76` to fix that.

**Comment by siom79** [comment]
Btw: A japicmp-plugin for Sonarqube sounds interesting. If you like, I can support you.

**Comment by LasneF** [comment]
ok after digging a little bit i ve isolated and found the reason 
because of transitive dep my project is using hk2-locator 
and looking on the pom here
http://grepcode.com/file/repo1.maven.org/maven2/org.glassfish.hk2/hk2-locator/2.1.88/META-INF/maven/org.glassfish.hk2/hk2-locator/pom.xml

it uses maven optional for  :'( 

this is our issue, the 2nd loop try to explore and to get the jar of this optionnal whereas in my case i do not have dowloaded it (because of Nexus pro policies) , and this make the rest of the analysis crashing .

The question is should the dependency resolver should follow optional ?   

looking on the doc here it should be excluded
https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html

**Comment by siom79** [comment]
Meanwhile I have also done some investigations and the second loop seems to necessary in order to resolve the dependencies added to the enclosing pom transitively. In my example I have created a maven module that has as only dependency another maven module which includes `com.fasterxml.jackson.core:jackson-databind:jar:2.6.3:compile` as dependency. The call `mavenParameters.getMavenProject().getArtifacts();` now only returns the dependency of the first project but not the dependency on jackson. The inner loop therefore resolves the first dependency and returns the three dependencies for the jackson databinding library.

When I now set the databind dependency on the second module as optional, then the maven resolution still returns it. The same is true when I declare the dependency of the enclosing module as optional. Hence maven does per default also resolve optional dependencies. 

To circumvent the resolution of optional dependencies, I have added a `ResolutionFilter` that does not include optional dependencies:
```
request.setResolutionFilter(new ArtifactFilter() {
	@Override
	public boolean include(Artifact artifact) {
		return !artifact.isOptional();
	}
});
```
Now dependencies declared to be optional are not resolved. I have committed that on the `develop` branch.
Let me know what you think.

**Comment by LasneF** [comment]
I ll on holiday for next 2 week. I will test this when be back
But this looks good.
Thx

**Comment by LasneF** [comment]
I have double check on my use case using the develop version  and now it is fine
thx for the fix

**Comment by siom79** [comment]
Thank you for your help. 

I will release it with 0.6.1.

**Comment by siom79** [comment]
Released with version 0.6.1.
